x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java

```diff
@@ -373,6 +373,103 @@ public void testMultiCountAllWithFilter() {
+
+        var limit = as(plan, LimitExec.class);
+        var exchange = as(limit.child(), ExchangeExec.class);
+        var project = as(exchange.child(), ProjectExec.class);
+        var field = as(project.child(), FieldExtractExec.class);
+        var query = as(field.child(), EsQueryExec.class);
+        assertThat(query.limit().fold(), is(1000));
+
+        Source filterSource = new Source(2, 37, "emp_no > 10000");
+        var range = wrapWithSingleQuery(queryText, QueryBuilders.rangeQuery("emp_no").gt(10010), "emp_no", filterSource);
+        var queryString = QueryBuilders.queryStringQuery("last_name: Smith");
+        var expected = QueryBuilders.boolQuery().must(queryString).must(range);
+        assertThat(query.query().toString(), is(expected.toString()));
+    }
+
+    /**
+     * Expecting
+     * LimitExec[1000[INTEGER]]
+     * \_ExchangeExec[[_meta_field{f}#1163, emp_no{f}#1157, first_name{f}#1158, gender{f}#1159, job{f}#1164, job.raw{f}#1165, langua
+     * ges{f}#1160, last_name{f}#1161, long_noidx{f}#1166, salary{f}#1162],false]
+     *   \_ProjectExec[[_meta_field{f}#1163, emp_no{f}#1157, first_name{f}#1158, gender{f}#1159, job{f}#1164, job.raw{f}#1165, langua
+     * ges{f}#1160, last_name{f}#1161, long_noidx{f}#1166, salary{f}#1162]]
+     *     \_FieldExtractExec[_meta_field{f}#1163, emp_no{f}#1157, first_name{f}#]
+     *       \_EsQueryExec[test], indexMode[standard],
+     *       query[{"bool":{"must":[{"query_string":{"query":"last_name: Smith","fields":[]}},
+     *       {"esql_single_value":{"field":"emp_no","next":{"range":{"emp_no":{"gt":10010,"boost":1.0}}},"source":"emp_no > 10010@3:9"}}],
+     *       "boost":1.0}}][_doc{f}#1167], limit[1000], sort[] estimatedRowSize[324]
+     */
+    public void testQueryStringFunctionMultipleWhereClauses() {
+        assumeTrue("skipping because QSTR_FUNCTION is not enabled", EsqlCapabilities.Cap.QSTR_FUNCTION.isEnabled());
+        String queryText = """
+            from test
+            | where qstr("last_name: Smith")
+            | where emp_no > 10010
+            """;
+        var plan = plannerOptimizer.plan(queryText, IS_SV_STATS);
+
+        var limit = as(plan, LimitExec.class);
+        var exchange = as(limit.child(), ExchangeExec.class);
+        var project = as(exchange.child(), ProjectExec.class);
+        var field = as(project.child(), FieldExtractExec.class);
+        var query = as(field.child(), EsQueryExec.class);
+        assertThat(query.limit().fold(), is(1000));
+
+        Source filterSource = new Source(3, 8, "emp_no > 10000");
+        var range = wrapWithSingleQuery(queryText, QueryBuilders.rangeQuery("emp_no").gt(10010), "emp_no", filterSource);
+        var queryString = QueryBuilders.queryStringQuery("last_name: Smith");
+        var expected = QueryBuilders.boolQuery().must(queryString).must(range);
+        assertThat(query.query().toString(), is(expected.toString()));
+    }
+
```

Comment:
Thanks @fang-xing-esql ! I've added these tests in c5b45ec0f43cd7b621ba306be90431093217eb2c.

Are there any other combinations that you can think of, or any rule of thumb that I could use to understand what cases should I be implementing? I haven't found more examples in `LocalPhysicalPlanOptimizerTests` that could guide me here.

>  it looks like multiple qstrs are allowed but not combined for now, perhaps it can be a candidate for further optimization.

Probably so - I wonder if queries pushed down to Lucene should be left to Lucene to optimize? 🤔

Reply:
Thanks for adding more tests!

> Are there any other combinations that you can think of, or any rule of thumb that I could use to understand what cases should I be implementing? I haven't found more examples in `LocalPhysicalPlanOptimizerTests` that could guide me here.

It could be challenging to write effective testcases. We always try to improve test coverage, but it is not guaranteed that everything is covered, there could be many edge cases, I found PR reviews help me a lot to improve test coverage. There are many existing test cases, and it is normal not be able to remember all of them :). These are the things that come to my mind first when writing tests for new functions:

1. Verify resolutions - function and type resolution. This is done(majority of them) by `Analyzer`,  and these tests are usually covered by `AnalyzerTests` and `VerifierTests`.
2. Verify optimizer builds an optimal/desired plan. These are usually covered by `LogicalPlanOptimizerTests`, `PhysicalPlanOptimizerTests`, `QueryTranslatorTests`, and their local version if exist. For the functions and operators that can appear under where clause, I would consider test compound predicates - conjunctive and disjunctive. For this function in particular, because it got pushed down into Lucene, we could get some combinations of bool, should, must queries, which makes me think that we may want to cover conjunctive and disjunctive, and perhaps some nested cases. Most of the ES|QL functions have their own compute engine implementation, and this doesn't, which makes me think that we may want to cover some combinations of them. I think these are not edge cases, they might be quite common use cases. I'll try to think about more scenarios, and update it here, just don't want to throw all of them at one time :-).
3. Verify the query returns correct results. These are mainly done in CsvTests, REST and yml tests. I would add more complicated(conjunctive and disjunctive) tests to verify we get correct results.

> 
> > it looks like multiple qstrs are allowed but not combined for now, perhaps it can be a candidate for further optimization.
> 
> Probably so - I wonder if queries pushed down to Lucene should be left to Lucene to optimize? 🤔

Yeah, Lucene might be able to do optimizations on multiple queries pushed down to it. Honestly I'm not an expert of it, still learning. There is something we can do in optimizer, an example of it is `CombineBinaryComparisons`, if Lucene doesn't do it or doesn't do it well, this rule can be a reference.